clodius/cli/aggregate.py

```diff
@@ -754,27 +841,159 @@ def line_to_np_array(line):
+
+    row = c.execute("SELECT * from tileset_info").fetchone()
+    version = row[next(zip(*c.description)).index("version")]
+    c.execute(
+        f"""
+        UPDATE tileset_info
+        SET version = '{version}t'
+        WHERE version = '{version}'
+        """
+    )
     conn.commit()
-    return True
+
+    c.execute(
+        """
+        CREATE TABLE tiles
+        (
+            id int,
+            intervalId int,
+            PRIMARY KEY (id, intervalId)
+        )
+        """
+    )
+
+    # I.e., tiles_cumsum[3] is the number of tiles with zoomlevels lower than 3
+    tiles_cumsum = np.cumsum([0] + [2 ** x for x in range(max_zoom + 1)])
+
+    interval_inserts = []
+    tile_inserts = []
+
+    def batch_insert(conn, c, interval_inserts, tile_inserts, counter):
+        if verbose > 0:
+            print(f"Insert batch ({counter})")
+
+        with transaction(conn):
+            c.executemany(
+                "INSERT INTO intervals VALUES (?,?,?,?,?,?,?,?,?)", interval_inserts
+            )
+            c.executemany("INSERT INTO tiles VALUES (?,?)", tile_inserts)
+
+        interval_inserts.clear()
+        tile_inserts.clear()
+
+    for interval_idx, interval in enumerate(sorted_intervals):
+        curr_zoom = 0
+        inserted = False
+        # go through each interval from most important to least
+        while curr_zoom <= max_zoom:
+            # try to place it in the highest zoom level and go down from there
+            tile_width = tile_size * 2 ** (max_zoom - curr_zoom)
+
+            curr_pos = interval[0]
+            space_available = True
+
+            # if we have not yet inserted the interval somewhere
+            if not inserted:
+                # check if there's space at this zoom level
+                while curr_pos < interval[1]:
+                    curr_tile = math.floor(curr_pos / tile_width)
+                    tile_id = f"{curr_zoom}.{curr_tile}"
+
+                    # if any of the overlapping tiles is already filled up,
+                    # lets go to the next zoom level by setting `space_available`
+                    # to false
+                    if tile_counts[tile_id] >= max_per_tile:
+                        space_available = False
+                        break
+
+                    curr_pos += tile_width
+
+            # if there is, then increment the tile counters
+            if not inserted and space_available:
+                curr_pos = interval[0]
+                while curr_pos < interval[1]:
+                    curr_tile = math.floor(curr_pos / tile_width)
+                    tile_id = "{}.{}".format(curr_zoom, curr_tile)
+
+                    tile_counts[tile_id] += 1
+
+                    curr_pos += tile_width
+
+            if inserted or space_available:
+                # there's available space
+                if not inserted:
+                    value = uid_to_interval[interval[-1]]
+                    interval_inserts.append(
```

Comment:
This part makes me a little suspicious. It looks like the entire interval could be inserted at multiple zoom levels. That may be necessary for the range-query based indexing but it shouldn't be necessary for the tile_id based indexing.

You should be able to insert each interval only once and then only reference it in the tile inserts.

Reply:
No, the interval is inserted only once. See the second `if` statement. Once `inserted == True` the interval won't be appended anymore.

The only thing that is copied are the tile-interval associations.